src/StoreApi/Authentication.php

```diff
@@ -20,15 +22,62 @@ public function init() {
 	 * @return \WP_Error|null|bool
 	 */
 	public function check_authentication( $result ) {
-		if ( ! empty( $result ) ) {
+		if ( ! $this->is_request_to_store_api() ) {
 			return $result;
 		}
 
-		if ( $this->is_request_to_store_api() ) {
-			return true;
+		$rate_limiting_options = RateLimits::get_options();
+
+		if ( $rate_limiting_options->enabled ) {
+			$action_id = 'store_api_request_';
+
+			if ( is_user_logged_in() ) {
+				$action_id .= get_current_user_id();
+			} else {
+				$ip_address = self::get_ip_address( $rate_limiting_options->proxy_support );
+
+				if ( ! $ip_address ) {
+					return new \WP_Error(
+						'ip_address_cannot_be_determined',
+						'Bad request. Client IP address cannot be determined.',
+						array( 'status' => 400 )
+					);
+				}
+
+				$action_id .= md5( $ip_address );
+			}
+
+			$retry  = RateLimits::is_exceeded_retry_after( $action_id );
+			$server = rest_get_server();
+			$server->send_header( 'RateLimit-Limit', $rate_limiting_options->limit );
+
+			if ( false !== $retry ) {
+				$server->send_header( 'RateLimit-Retry-After', $retry );
+				$server->send_header( 'RateLimit-Remaining', 0 );
+				$server->send_header( 'RateLimit-Reset', time() + $retry );
+
+				$ip_address = $ip_address ?? self::get_ip_address( $rate_limiting_options->proxy_support );
+				if ( $ip_address ) {
+					do_action( 'woocommerce_store_api_rate_limit_exceeded', $ip_address );
```

Comment:
Why don't we fire this regardless of IP?

Reply:
Good call@mikejolley, we were only thinking about this for logging offending IPs, and totally forgot that logged-in Users might also be suspects :high-five: 
Removing the `if`.